### ISSUE TITLE
Use Builtin Higher Res Fonts

### ISSUE THREAD

**Issue by bangstk** [fgsfdsfgs/perfect_dark#211]
In my opinion, the low res main font is the most jarring part of the game visually.

The game has two higher res versions of the font built in already (which it uses in the credits and the opening copyright text screen for instance) which should look a lot better even with filtering enabled. Could it be possible to have the game override the main font (used in menus, briefings, HUD messages, etc) with the highest res builtin font?

This would be a fully stock assets alternative to using 1964GEPD's third party UI texture replacements. (though that would also be a nice feature)

Based on my messing with the rom in the past I know the higher res builtin charsets should be complete, though some of the UV mapping is incorrect (for 'W' character for instance) but that should all be fixable.

I think the ammo number/timer digit graphics should be left alone for style reasons but everything else should be fair game.

**Comment by fgsfdsfgs** [comment]
I'm not sure if, for example, the font used in the legal screen is exactly the same (but higher res) as the one used in the HUD messages. I suppose this is theoretically possible to implement, but it would probably be easier to add HD texture support and just use the fonts that ship with 1964GEPD.

**Comment by Xii-Nyth** [comment]
different mipmaps also have different outlines, so its would be best just to use the hires replacements when we get tetxure replacement.

**Comment by Xii-Nyth** [comment]
> I'm not sure if, for example, the font used in the legal screen is exactly the same (but higher res) as the one used in the HUD messages. I suppose this is theoretically possible to implement, but it would probably be easier to add HD texture support and just use the fonts that ship with 1964GEPD.

when we get texture replacement please contact me asap!! thanks

**Comment by amazingalek** [comment]
> In my opinion, the low res main font is the most jarring part of the game visually.

I agree! Everything else looks amazing, the font really stands out. Higher res font is my most anticipated feature :)

>I think the ammo number/timer digit graphics should be left alone for style reasons but everything else should be fair game.

Interesting, to me it makes sense to increase the res of those too, why do you think they should stay as they are?

**Comment by amazingalek** [comment]
I've been looking into this. I wanted to confirm that there are indeed separate fonts being used, and see what happened if I swap them. This is what I've found so far:

The legal text is rendered in title.c. Comparing to the text rendered in mainmenu.c, it's done the exact same way. Same fonts and parameters to the render function (textRenderProjected).

There are three fonts used in title and main menu: small (g_CharsHandelGothicSm/g_FontHandelGothicSm), medium (g_CharsHandelGothicMd/g_FontHandelGothicMd) and large (g_CharsHandelGothicLg/g_FontHandelGothicLg). Changing the font anywhere also changes the text size, there's no separate scaling parameter I can find.

There's a textReset function in game_1531a0.c which at first glance seemed promising. It swaps out fonts globally based on which screen you're on. But this seems like a red herring, the same fonts seem to be used anyway.

My working theory is that the difference between the title screen and menus has nothing to do with the fonts directly, but that these screens are rendered with different resolutions, or something similar, which affects text scaling/quality.

I've been experimenting with different resolutions etc. but it's way outside my expertise, so I'm just doing shots in the (perfect) dark. No luck so far. Maybe a render guy/gal could chime in? :-)

**Comment by telemarker** [comment]
any news on this - i think the blury text in menus and in game popup is the only annoying thing from the port (well the pink line in the beginning too)...

**Comment by amazingalek** [comment]
I did figure out how to make the fonts crispy. No blur, very pixelated, still low res. Not sure if it's an improvement or not. Can post some images tonight.

**Comment by telemarker** [comment]
awesome thanks and please let us know how to update the existing install.

**Comment by amazingalek** [comment]
Here's the crispy font:
![pd_menu_point](https://github.com/user-attachments/assets/e635c629-ff0e-497d-9083-bce21526bb1f)

For comparison, this is before the change:
![pd_menu_bilerp](https://github.com/user-attachments/assets/4d5f82d4-e571-4a12-870f-a1f9dc565905)

All I did was add this line at the top of the `textRenderProjected` function in `game_1531a0.c`:
`gDPSetTextureFilter(gdl++, G_TF_POINT);`

It seems to make all text in the game crispy.
Do you think it's an improvement?

**Comment by telemarker** [comment]
it is a tiny improvement but it seems you have to decompile something to achieve it - i cant see any game_1531a0.c - either the root nor the data folder. maybe its possible to set it in the pd.ini?
thanks

**Comment by fgsfdsfgs** [comment]
> All I did was add this line at the top of the textRenderProjected function in game_1531a0.c:
> gDPSetTextureFilter(gdl++, G_TF_POINT);

This is already an option in the game. Extended -> Video -> disable GUI Texture Filtering and restart.

**Comment by telemarker** [comment]
thanks, how is the progress of implementing real hires fonts?

**Comment by fgsfdsfgs** [comment]
No progress yet.

**Comment by Graslu** [comment]
I've been given permission to release the raw PNG files of the HD HUD texture packs of GE and PD here: https://github.com/Graslu/GEPDTexturePacks

Feel free to use these in the future if / when HD texture replacements are possible.

**Comment by telemarker** [comment]
hey graslu thanks thats a great released, but it looks there is no way to install it on the fork right now?

**Comment by Graslu** [comment]
That's correct. This is just the raw files from the 1964 GEPD HQ pack that can be used with emulators. It is currently not possible to use with the PC port.

**Comment by bangstk** [comment]
I'm actually the Trov who made that PD hi res font pack.

I know that the original game loads all of the letters from a single texture (for each font). I think it would make most sense to make a hi res version of those, rather than trying to get the game to load them the way the emulator version is set up (separate textures for each letter). Additionally some stuff like the glowy-outline version of the fonts I'm pretty sure are dynamically produced by the game and would be redundant in the texture pack.

I suppose I can come up with the single-texture versions of the hi res fonts for anyone who feels like experimenting with it. I will make sure all the letter positions match so that the original UV mapping logic can probably be reused (other than scaling for the larger base texture)
